repl/src/main/scala/com/cloudera/livy/repl/ReplDriver.scala

```diff
@@ -73,19 +63,28 @@ class ReplDriver(conf: SparkConf, livyConf: RSCConf)
     }
   }
 
-  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.ReplJobRequest): Unit = {
-    Future {
-      jobFutures(msg.id) = session.execute(msg.code).result
-    }
+  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.ReplJobRequest): Int = {
+    session.execute(msg.code)
   }
 
-  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResult): String = {
-    val result = jobFutures.getOrElse(msg.id, null)
-    Option(result).map { r => compact(render(r)) }.orNull
+  /**
+    * Return statement results. Results are sorted by statement id.
+    */
+  def handle(ctx: ChannelHandlerContext, msg: BaseProtocol.GetReplJobResults): ReplJobResults = {
+    val stmts = if (msg.allResults) {
+      session.statements.values.toArray
+    } else {
+      assert(msg.from != null)
+      assert(msg.size != null)
+      val until = msg.from + msg.size
+      session.statements.filterKeys(id => id >= msg.from && id < until).values.toArray
```

Comment:
I think there're two potential issues here:

1. Since `statements` is mutable, here without synchronization, the returned result may be inconsistent.
2. Traverse through the who map will be slower if Livy is running for hours, it would be better to use other data structure like sorted map instead.

Also I'm not sure in which scenario do we require to bulk get the statement results?